targets/simple_switch/sample_extern/macsec/crypto.cpp

```diff
@@ -0,0 +1,470 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Antonin Bas (antonin@barefootnetworks.com)
+ * 
+ * Modifications:
+ * Marco Haeberle (marco.haeberle@uni-tuebingen.de)
+ * Joshua Hartmann
+ *
+ */
+
+// aes stuff
+#include <openssl/aes.h>
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+
+#include <bm/bm_sim/_assert.h>
+#include <bm/bm_sim/parser.h>
+#include <bm/bm_sim/tables.h>
+#include <bm/bm_sim/logger.h>
+// needed for externs
+#include <bm/bm_sim/extern.h>
+#include <unistd.h>
+
+#include <condition_variable>
+#include <string>
+#include <vector>
+#include <algorithm>
+#include <iostream>
+#include <deque>
+#include <fstream>
+#include <sstream>
+#include <mutex>
+
+#include "simple_switch.h"
+#include "register_access.h"
+
+using std::endl;
+using std::cout;
+using std::string;
+using std::vector;
+
+#define SAK_SIZE 16
+#define SCI_SIZE 8
+#define PN_SIZE 4
+#define ADDR_SIZE 6
+#define SECTAG_SIZE 16
+#define ICV_SIZE 16
+#define IPV4_HDR_SIZE 20
+#define ETHERTYPE_SIZE 2
+
+namespace bm {
+class ExternCrypt : public ExternType {
+ public:
+  BM_EXTERN_ATTRIBUTES {
+  }
+
+  void init() override {
+  }
+  void protect(const Data &in_sak,
+               const Data &in_sci,
+               const Data &in_pn,
+               const Data &in_src_addr,
+               const Data &in_dst_addr,
+               const Data &in_sectag,
+               const Data &in_ethertype,
+               const Data &in_prepend_ipv4_hdr,
+               const Data &in_ipv4_hdr) {
+    std::vector<unsigned char> secure_association_key = get_char_vector(
+                                          in_sak.get_string(), SAK_SIZE);
+    std::vector<unsigned char> secure_channel_identifier =
+      get_char_vector(in_sci.get_string(), SCI_SIZE);
+
+    std::vector<unsigned char> packet_number =
+        get_char_vector(in_pn.get_string(), PN_SIZE);
+
+    std::vector<unsigned char> source_mac_address =
+      get_char_vector(in_src_addr.get_string(), ADDR_SIZE);
+
+    std::vector<unsigned char> destination_mac_address =
+      get_char_vector(in_dst_addr.get_string(), ADDR_SIZE);
+
+    std::vector<unsigned char> security_tag =
+      get_char_vector(in_sectag.get_string(), SECTAG_SIZE);
+
+    std::vector<unsigned char> ethertype =
+      get_char_vector(in_ethertype.get_string(), ETHERTYPE_SIZE);
+
+    bool prepend_ipv4 = false;
+    // must pass byte to external function
+    // use 0x54 T as true
+    // use 0x46 F as false
+    // cout << "[p4sec] prepend IPv4 Header ? "
```

Comment:
did you mean to keep this?

Reply:
I will remove the comments and keep the prepend code.  Note, the extern has a protect method with one arg as prepend_ipv4_hdr.  This is why the prepend code has to be retained.